### PR TITLE
Better mixed runtime & function deploy handling

### LINKED ISSUE #161: Fatal error when trying to deploy single-functions
**FYI**: This is with the latest code from master
Version: serverless@1.26.1
Steps to reproduce: Create any/simple stack which uses this plugin, and try to do a single-function deploy.
```
  Error --------------------------------------------------
 
  Unsupported type: data is neither a path or a Buffer
 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
 
  Stack Trace --------------------------------------------
 
Error: Unsupported type: data is neither a path or a Buffer
    at Object.ZipLocal.sync.unzip (/Users/farley/myService/node_modules/zip-local/main.js:386:15)
    at injectRequirements (/Users/farley/myService/node_modules/serverless-python-requirements/lib/inject.js:45:27)
    at ServerlessPythonRequirements.injectAllRequirements (/Users/farley/myService/node_modules/serverless-python-requirements/lib/inject.js:118:5)
From previous event:
    at PluginManager.invoke (/Users/farley/.npm-packages/lib/node_modules/serverless/lib/classes/PluginManager.js:372:22)
    at PluginManager.run (/Users/farley/.npm-packages/lib/node_modules/serverless/lib/classes/PluginManager.js:403:17)
    at variables.populateService.then (/Users/farley/.npm-packages/lib/node_modules/serverless/lib/Serverless.js:102:33)
    at runCallback (timers.js:672:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)
From previous event:
    at Serverless.run (/Users/farley/.npm-packages/lib/node_modules/serverless/lib/Serverless.js:89:74)
    at serverless.init.then (/Users/farley/.npm-packages/lib/node_modules/serverless/bin/serverless:42:50)
```

```
# serverless.yml
service: myService

provider:
  name: aws
  runtime: python2.7

plugins:
  - serverless-python-requirements

custom: 
  pythonRequirements:
    dockerizePip: true

functions:
  hello:
    handler: handler.hello
```

```
# requirements.txt
requests
```

master/bleeding edge might not be ready yet, I was just developing something on top of it and ran into this.

Cheers!

### LINKED ISSUE #179: Prevent plugin packaging when deploying non-python functions
It would be nice if the plugin could either ignore or not be invoked when deploying a non-python function. For those of us with mixed-runtime projects, this will speed up testing and deployment. 

Right now, when deploying a node function, all the python packages we have defined for our python packages get included in the node function's ZIP.

I've had to define our provider-level runtime to python3.6 and then at the function-level I repeat the runtime definitions (python3.6, node6.10, etc). Here's an example:

```yaml
service: my-mixed-runtime-service

plugins:
  - serverless-python-requirements
custom:
  pythonRequirements:
    dockerizePip: non-linux
    fileName: requirements.txt

provider:
  name: aws
  runtime: python3.6
  stage: test
  region: us-east-1

functions:
  node-function:
    handler: functions/my_node_function.handler
    runtime: nodejs6.10
    timeout: 10
  python-function:
    handler: functions/my_python_function.handler
    runtime: python3.6
    timeout: 10
    package: 
      excludes: ['node_modules/']
```

Then, when running a `sls deploy --function node-function`, I get the following output. 42MB is accurate for our python packages but we would expect our node_modules to be around 9-10MB.
```
Serverless: Installing required Python packages with python3.6...
Serverless: Docker Image: lambci/lambda:build-python3.6
Serverless: Linking required Python packages...
Serverless: Packaging function: node-function...
Serverless: Excluding development dependencies...
Serverless: Unlinking required Python packages...
Serverless: Uploading function: node-function (42.49 MB)...
Serverless: Successfully deployed function: node-function
Serverless: Successfully updated function: node-function
```

And here's our requirements.txt and package.json for good measure:

**Requirements.txt**
```
appdirs==1.4.3
cached-property==1.4.0
certifi==2018.1.18
chardet==3.0.4
defusedxml==0.5.0
idna==2.6
isodate==0.6.0
lxml==4.2.1
pytz==2018.3
requests==2.18.4
requests-toolbelt==0.8.0
six==1.11.0
urllib3==1.22
zeep==2.5.0
```
**Package.json**
```json
{
  "name": "my-mixed-runtime-service",
  "version": "1.0.0",
  "description": "Blah",
  "main": "README.md",
  "dependencies": {
    "@slack/client": "^4.1.0",
    "async": "^2.6.0",
    "csv-parse": "^2.2.0",
    "detect-csv": "^1.1.0",
    "moment": "^2.22.0",
    "pg": "^7.4.1",
    "request": "^2.85.0",
    "request-promise": "^4.2.2",
    "serverless-pseudo-parameters": "^1.4.2",
    "serverless-python-requirements": "^3.3.1",
    "serverless-step-functions": "^1.4.0"
  },
  "devDependencies": {},
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "repository": {
    "type": "git",
    "url": "blah"
  },
  "author": "",
  "license": "ISC",
  "homepage": "blah"
}

```

### PR DESCRIPTION
**Pull request by dschep**
fixes #161 and fixes #179 

Could you test this out @icj217 & @AndrewFarley?

```
npm i UnitedIncome/serverless-python-requirements#issues-161-179
```

### CODE REVIEW

**Comment by icj217** [comment]
@dschep Tried it out and here's what I get (`historical-validation` and `convert-ens-format` are both python functions)
Function definition:
```yaml
custom  
   pythonRequirements:
    dockerizePip: 'non-linux'

package:
  individually: true
  excludeDevDependencies: false

  historical-validation:
    handler: historical-validation.handler (also tried functions/historical-validation/historical-validation.handler)
    runtime: python3.6
    description: Historical validator
    timeout: 300
    module: functions/historical-validation
```
Deployment:
```
craigburdulis@Craig-Burdulis-MacBook-Air:~/Documents/Projects/hsx-panel-qa$ SLS_DEBUG=* sls deploy --function historical-validation
Serverless: Load command run
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command create
Serverless: Load command install
Serverless: Load command package
Serverless: Load command deploy
Serverless: Load command deploy:function
Serverless: Load command deploy:list
Serverless: Load command deploy:list:functions
Serverless: Load command invoke
Serverless: Load command invoke:local
Serverless: Load command info
Serverless: Load command logs
Serverless: Load command login
Serverless: Load command logout
Serverless: Load command metrics
Serverless: Load command print
Serverless: Load command remove
Serverless: Load command rollback
Serverless: Load command rollback:function
Serverless: Load command slstats
Serverless: Load command plugin
Serverless: Load command plugin
Serverless: Load command plugin:install
Serverless: Load command plugin
Serverless: Load command plugin:uninstall
Serverless: Load command plugin
Serverless: Load command plugin:list
Serverless: Load command plugin
Serverless: Load command plugin:search
Serverless: Load command emit
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command rollback
Serverless: Load command rollback:function
Serverless: Load command invoke
Serverless: Load command invoke:stepf
Serverless: Load command requirements
Serverless: Load command requirements:clean
Serverless: Load command requirements:install
Serverless: Invoke deploy
Serverless: Invoke deploy:function
Serverless: Installing requirements of requirements.txt in .serverless...
Serverless: Docker Image: lambci/lambda:build-python3.6
Serverless: Installing requirements of functions/historical-validation/requirements.txt in .serverless/functions/historical-validation...
Serverless: Docker Image: lambci/lambda:build-python3.6
Serverless: Installing requirements of functions/convert-ens-format/requirements.txt in .serverless/functions/convert-ens-format...
Serverless: Docker Image: lambci/lambda:build-python3.6
Serverless: Invoke package:function
Serverless: Packaging function: historical-validation...
Serverless: /Users/craigburdulis/Documents/Projects/hsx-panel-qa/.serverless/historical-validation.zip
Serverless: Injecting required Python packages to package...

  Type Error ---------------------------------------------

  Cannot read property 'artifact' of undefined

     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.

  Stack Trace --------------------------------------------

TypeError: Cannot read property 'artifact' of undefined
    at BbPromise.resolve.filter.map.map.func (/Users/craigburdulis/Documents/Projects/hsx-panel-qa/node_modules/serverless-python-requirements/lib/inject.js:98:40)
From previous event:
    at PluginManager.invoke (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:372:22)
    at PluginManager.spawn (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:390:17)
    at Deploy.BbPromise.bind.then.then (/usr/local/lib/node_modules/serverless/lib/plugins/deploy/deploy.js:118:50)
From previous event:
    at Object.before:deploy:deploy [as hook] (/usr/local/lib/node_modules/serverless/lib/plugins/deploy/deploy.js:113:10)
    at BbPromise.reduce (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:372:55)
From previous event:
    at PluginManager.invoke (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:372:22)
    at PluginManager.run (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:403:17)
    at variables.populateService.then (/usr/local/lib/node_modules/serverless/lib/Serverless.js:102:33)
    at runCallback (timers.js:672:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)
From previous event:
    at Serverless.run (/usr/local/lib/node_modules/serverless/lib/Serverless.js:89:74)
    at serverless.init.then (/usr/local/lib/node_modules/serverless/bin/serverless:42:50)

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless

  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           6.11.0
     Serverless Version:     1.26.1
```

**Comment by icj217** [comment]
It does work, however, when I set `individually` to false.

**Comment by dschep** [comment]
yup. good catch. i'm able to reproduce the issue.

**Comment by icj217** [comment]
Deploying node.js functions now works as well when `individually` set to false.

**Comment by AndrewFarley** [comment]
Checked this against three of my codebases with a variety of configs, worked fine, passed all my codebases' CI-related tests.  Glanced at the code also, seems good.  Much appreciated.  Before merging I would remove that erroneous console.log of artifacts.

Of note: None of mine had the error like yours above with or without individually packaging.  Might be a difference in the version of NodeJS (I'm using 6.13), but it should be pretty easy to fix (checking typeof, or so).

**Comment by dschep** [comment]
Updated. Give it another shot @icj217. Glad to hear it works for you @AndrewFarley!

File: lib/inject.js [hunk 1]
```diff
@@ -107,14 +107,14 @@ function injectAllRequirements() {
       .map(func =>
         injectRequirements(
           path.join('.serverless', func.module, 'requirements'),
-          func.package.artifact,
+          func.package ? func.package.artifact : artifact,
```
[the remark this answers is not in the document]
Comment (dschep) [reply lib/inject.js]: updated line 97 too.

**Comment by icj217** [comment]
Now it's saying:
```
ReferenceError: artifact is not defined
    at BbPromise.resolve.filter.map.map.func (/Users/craigburdulis/Documents/Projects/hsx-panel-qa/node_modules/serverless-python-requirements/lib/inject.js:97:67)
From previous event:
    at PluginManager.invoke (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:372:22)
    at PluginManager.spawn (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:390:17)
    at Deploy.BbPromise.bind.then.then (/usr/local/lib/node_modules/serverless/lib/plugins/deploy/deploy.js:118:50)
From previous event:
    at Object.before:deploy:deploy [as hook] (/usr/local/lib/node_modules/serverless/lib/plugins/deploy/deploy.js:113:10)
    at BbPromise.reduce (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:372:55)
From previous event:
    at PluginManager.invoke (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:372:22)
    at PluginManager.run (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:403:17)
    at variables.populateService.then (/usr/local/lib/node_modules/serverless/lib/Serverless.js:102:33)
    at runCallback (timers.js:672:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)
From previous event:
    at Serverless.run (/usr/local/lib/node_modules/serverless/lib/Serverless.js:89:74)
    at serverless.init.then (/usr/local/lib/node_modules/serverless/bin/serverless:42:50)
```

**Comment by icj217** [comment]
@dschep Now its line 99. Seems like the `func.package` does not exist yet when doing individual packaging.

```
TypeError: Cannot set property 'artifact' of undefined
    at BbPromise.resolve.filter.map.map.func (/Users/craigburdulis/Documents/Projects/hsx-panel-qa/node_modules/serverless-python-requirements/lib/inject.js:99:33)
From previous event:
    at PluginManager.invoke (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:372:22)
    at PluginManager.spawn (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:390:17)
    at Deploy.BbPromise.bind.then.then (/usr/local/lib/node_modules/serverless/lib/plugins/deploy/deploy.js:118:50)
From previous event:
    at Object.before:deploy:deploy [as hook] (/usr/local/lib/node_modules/serverless/lib/plugins/deploy/deploy.js:113:10)
    at BbPromise.reduce (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:372:55)
From previous event:
    at PluginManager.invoke (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:372:22)
    at PluginManager.run (/usr/local/lib/node_modules/serverless/lib/classes/PluginManager.js:403:17)
    at variables.populateService.then (/usr/local/lib/node_modules/serverless/lib/Serverless.js:102:33)
    at runCallback (timers.js:672:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)
From previous event:
    at Serverless.run (/usr/local/lib/node_modules/serverless/lib/Serverless.js:89:74)
    at serverless.init.then (/usr/local/lib/node_modules/serverless/bin/serverless:42:50)
```

**Comment by dschep** [comment]
Hmm. weird. I was able to deploy one of the test services now when i try the following I am at getting the errors @AndrewFarley reported.
```
git clone https://github.com/UnitedIncome/serverless-python-requirements
cd serverless-python-requirements
git checkout issues-161-179
cd tests/base
npm i $(npm pack ../..)
sls deploy --individually
sls deploy --individually -f hello
sls deploy --individually -f hello3
```

**Comment by dschep** [comment]
Hmm.... @AndrewFarley `sls deploy --individually -f hello` after doing `rm -r .serverless`

**Comment by dschep** [comment]
sorry. nope. haven't had time to prioritize on this since we don't use this feature ourselves.

**Comment by dschep** [comment]
@AndrewFarley i merged #181 into this PR since i think it should address the corrupted zip issue

**Comment by AndrewFarley** [comment]
@dschep Still got a problem when trying to deploy a single function with `serverless deploy function -f funcname` so can't validate if that fixes the corrupt zip bug.  No other deploy problems when not deploying single functions though.  See below...

```
Serverless: Excluding development dependencies...
Serverless: Injecting required Python packages to package...
 
  Type Error ---------------------------------------------
 
  path must be a string or Buffer
 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
 
  Stack Trace --------------------------------------------
 
TypeError: path must be a string or Buffer
    at TypeError (native)
    at fs.readFile (fs.js:305:11)
    at go$readFile (/Users/farley/PROJECTNAMEHERE/api_v2/node_modules/graceful-fs/graceful-fs.js:73:14)
    at Object.readFile (/Users/farley/PROJECTNAMEHERE/api_v2/node_modules/graceful-fs/graceful-fs.js:70:12)
    at Object.readFile (/Users/farley/PROJECTNAMEHERE/api_v2/node_modules/universalify/index.js:5:67)
    at injectRequirements (/Users/farley/PROJECTNAMEHERE/api_v2/node_modules/serverless-python-requirements/lib/inject.js:24:6)
    at ServerlessPythonRequirements.injectAllRequirements (/Users/farley/PROJECTNAMEHERE/api_v2/node_modules/serverless-python-requirements/lib/inject.js:118:12)
    at ServerlessPythonRequirements.BbPromise.bind.then.then (/Users/farley/PROJECTNAMEHERE/api_v2/node_modules/serverless-python-requirements/index.js:135:43)
    at runCallback (timers.js:672:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless
 
  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           6.12.3
     Serverless Version:     1.26.0
```

**Comment by icj217** [comment]
@dschep Seeing the same thing as @AndrewFarley

**Comment by AndrewFarley** [comment]
@dschep Isn't this still the upstream problem of the artifact not being set properly...?
```
  package: 
   { individually: undefined,
     path: undefined,
     artifact: undefined,
```

**Comment by dschep** [comment]
You are correct that package.artifact doesn't get set, but the packageFunction hook is passed the function object which should have an artifact set on it. https://github.com/UnitedIncome/serverless-python-requirements/pull/180/files#diff-168726dbe96b3ce427e7fedce31bb0bcR136

**Comment by AndrewFarley** [comment]
@dschep I fixed the problem with the following at the end of my file... this seems to fix all the problems.  This was in the `if this.serverless.service.package.individually` stanza above, worked perfectly below also.
```
    return injectRequirements(
      path.join('.serverless', 'requirements'),
      this.serverless.service.package.artifact ? this.serverless.service.package.artifact : funcArtifact,
      this.options
    );
```

**Comment by dschep** [comment]
Huh. Thought i'd done that for some reason. Added it. Thanks! :+1:

**Comment by AndrewFarley** [comment]
Super excited to get single function deploys working again!  :)  I had a huge grin on my face just now seeing it work again, will (re)accelerate developer velocity over here.  Cheers!

File: lib/inject.js [hunk 2]
```diff
@@ -117,7 +117,7 @@ function injectAllRequirements(funcArtifact) {
   } else {
     return injectRequirements(
       path.join('.serverless', 'requirements'),
-      this.serverless.service.package.artifact,
+      this.serverless.service.package.artifact || funcArtifact
```
Comment (icj217) [inline lib/inject.js]: Getting a syntax error. I think you're missing the `,` at end of line.

Comment (dschep) [reply lib/inject.js]: Ah. yes! (and that's probably why depcheck didn't think we use glob-all too)

**Comment by dschep** [comment]
I'll go ahead and merge it (but not yet push to npm) and test it out on my personal projects tonight. Yeah, a known set of contributors/users who are willing to test features before release would be nice to have. I'll consider you as being on that list for now :wink:

**Comment by cgrimal** [comment]
Hi guys, I was following the issue, especially after I updated the plugin before deploying an update of my Lambda function and it did not work!

I ended up pulling the branch I forked to implement the per function requirements in order to be able to deploy, but now that I have a bot more time, I pulled master, and I'm afraid I still have the `Cannot set property 'artifact' of undefined` error when deploying a single function.

I'm gonna try to have a look at the code and I'll keep you updated.

**Comment by cgrimal** [comment]
The problem seems to be that I have 2 functions defined in my project, and when I try to deploy the first one, the whole process (requirements installation, etc.) is done for both function. I don't think this is a recent issue, but now, the artifact property is not defined for the functions other than the one I want to deploy, hence the error.

@AndrewFarley: you say it works for you, so I'm curious, do you have one function or multiple functions defined in your serverless.yml file? If you only have one, can you try adding another one and try again?

If I'm correct, one solution would be to check what function is being processed. In almost all functions, we do something like:
```
if this.serverless.service.package.individually:
    for each function do:
       // do stuff
```
when it should be:
```
if this.serverless.service.package.individually:
    for each function do:
        if not single function deploy or function is the one deployed:
           // do stuff
```

Does it make sense?

**Comment by dschep** [comment]
That makes sense @cgrimal! Feel free to make a PR if you have time before I do.

**Comment by cgrimal** [comment]
I think I manage to have a "working" version, but the injection part was very long, and the vendors were not included. probably unrelated but not workable for me...

This is the end of the day in my timezone and I'll be on annual leave tomorrow night, so I'm afraid I'll have to leave it to you guys, sorry.

Here are the modifications I did in inject.js:
- import `lodash.keys` instead of `values`
- replace the loop on `values(this.serverless.service.functions)` to `keys(this.serverless.service.functions)`
- add this filter `.filter(f => f == this.serverless.variables.options.function)`
- update the previous filter to `.filter(f => (this.serverless.service.functions[f].runtime || this.serverless.service.provider.runtime).match(/^oython.*/))`
- in the first `map`, get the func body with: `var func = this.serverless.service.functions[f];` (f being the key)

So it is very basic, I just iterate over the keys instead of the values, as this is the key being passed as CLI argument to deploy a single version, that I get from `this.serverless.variables.options.function`.

The same should be done in other modules.

I hope it helps, really sorry to leave it like that. If there are still issues when I'm back in mid-May, I'll have a look!

Many thanks

**Comment by dschep** [comment]
No worries. Thanks for the notes. Enjoy your vacation!

**Comment by AndrewFarley** [comment]
Re: Most of my stacks have multiple functions, and I injected one of a non-python runtime for some brief testing and didn't run into any issues, with or without individually set.  I'm sure I didn't try out every single usage configuration path I'm sure though.  Probably need more tests to help cover/catch all these things that seem to not get caught currently.

**Comment by dschep** [comment]
I'm gonna try to add a second set of tests (`deploy-tests.bats` probably) but wont run them in CI since I don't want people to be able to deploy random crap to my AWS account by changing the test lambdas and making a PR. Test cases we need AFAICT:

* basic python service
* python service with individually & a non-python function
  * deploy
  * deploy -f py_func
  * deploy -f nodeFunc
* node service with a python function with individually

Anyother things i shoudl consider?

(i'm even gonna also actually invoke the lambdas, which will prove that the packaged deps are really usable too!)

**Comment by AndrewFarley** [comment]
@dschep Great plan, I think that mostly covers it... I agree don't run those in CI, but definitely make them something you/we can run locally to check/guarantee(ish) that things that are generated are sane.  Definitely make sure to try out the lambdas, make sure (of course) when you try the lambdas, you're including trying out using a _super_ simple pip module as well, to catch incase it doesn't zip deps.

**Pull request merged**